cli/tests/test_parallelize_command.rs

```diff
@@ -0,0 +1,181 @@
+    ◉  dc0e5d6135ce 1
+    ◉  000000000000
+    "###);
+
+    test_env.jj_cmd_ok(&workspace_path, &["parallelize", "dc0::"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @  6c7b60a45eb6 6
+    │ ◉  296f48966777 5
+    ├─╯
+    │ ◉  524062469789 4
+    ├─╯
+    │ ◉  a9334ecaa379 3
+    ├─╯
+    │ ◉  3a7b37ebe843 2
+    ├─╯
+    │ ◉  761e67df44b7 1
+    ├─╯
+    ◉  000000000000
+    "###);
+}
+
+// Only the head commit has descendants.
+#[test]
+fn test_parallelize_with_descendants_simple() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let workspace_path = test_env.env_root().join("repo");
+
+    for n in 1..6 {
+        test_env.jj_cmd_ok(&workspace_path, &["commit", &format!("-m{n}")]);
+    }
+    test_env.jj_cmd_ok(&workspace_path, &["describe", "-m=6"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @  b911505e443e 6
+    ◉  2e00cb15c7b6 5
+    ◉  9df3c87db1a2 4
+    ◉  9f5b59fa4622 3
+    ◉  d826910d21fb 2
+    ◉  dc0e5d6135ce 1
+    ◉  000000000000
+    "###);
+
+    test_env.jj_cmd_ok(&workspace_path, &["parallelize", "dc0::9df"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @  f28f986c7134 6
+    ◉        21e9963ac5ff 5
+    ├─┬─┬─╮
+    │ │ │ ◉  524062469789 4
+    │ │ ◉ │  a9334ecaa379 3
+    │ │ ├─╯
+    │ ◉ │  3a7b37ebe843 2
+    │ ├─╯
+    ◉ │  761e67df44b7 1
+    ├─╯
+    ◉  000000000000
+    "###);
+}
+
+// More than one commit being parallelized has descendants, but only the
+// descendants of the head commit will be rebased onto the target commits.
+#[test]
+fn test_parallelize_with_descendants_complex() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let workspace_path = test_env.env_root().join("repo");
+
+    for n in 1..6 {
+        test_env.jj_cmd_ok(&workspace_path, &["commit", &format!("-m{n}")]);
+    }
+    test_env.jj_cmd_ok(&workspace_path, &["new", "d82", "-m=child-of-2"]);
+    test_env.jj_cmd_ok(&workspace_path, &["new", "2e0", "-m=6"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @  d27ee705f7a9 6
+    ◉  2e00cb15c7b6 5
+    ◉  9df3c87db1a2 4
+    ◉  9f5b59fa4622 3
+    │ ◉  60c93e41d09c child-of-2
+    ├─╯
+    ◉  d826910d21fb 2
+    ◉  dc0e5d6135ce 1
+    ◉  000000000000
+    "###);
+
+    test_env.jj_cmd_ok(&workspace_path, &["parallelize", "dc0::9df"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @  9f1bec0d6c46 6
+    ◉        7dd2f5648395 5
+    ├─┬─┬─╮
+    │ │ │ ◉  b8f977c12383 4
+    │ │ ◉ │  7be8374575b9 3
+    │ │ ├─╯
+    ◉ │ │  2bfe3fe3e472 1
+    ├───╯
+    │ │ ◉  bf7eef1f1dc6 child-of-2
+    │ ├─╯
+    │ ◉  96ce11389312 2
+    ├─╯
+    ◉  000000000000
+    "###);
+}
+
+#[test]
+fn test_parallelize_failure_multiple_heads() {
```

Comment:
Would be good to have another test case that tries to parallelize a linear chain without specifying all commits in the chain (e.g. `jj parallelize <commit 1> <commit 3>`). I suppose that should be an error.

Reply:
Yeah good point. Maybe it should return an error, but it could also just make the commits siblings as long as the descendants are rebased correctly. I pushed a `test_parallelize_failure_disjoint_target_commits` that shows what happens currently. 

Let me fix the `test_parallelize_with_descendants_complex` case first and then we can come back to this and see if we want to return an error or allow it.

Reply:
Yes, I do think it's related to the case we talked about above. That's actually why I concluded it should be an error. My thinking was that what the user expresses with `jj parallelize` is that the target commits are actually independent (i.e. contrary to what the shape of the graph before they ran the command indicated). So if they're saying that commit 1 and commit 3 are independent, then they're saying that commit 3 can be made a sibling of commit 1, but we still assume that the descendants of the target commits maintain their dependencies, so commit 2 remains a child of commit 1, and commit 3 remains a child of commit 2, which is a contradiction of what the user expressed with `jj parallelize <commit 1> <commit 2>`.

Reply:
Yeah, I think you're right that it should be an error. I pushed a WIP commit that handles this case by returning an error. I probably need to get a little more clever about how I'm using the revset APIs. I'm going to take another look at it, but maybe you and Yuya have some suggestions.